File: src/php/lib/Grpc/CallInvoker.php
<?php
/*
 *
 * Copyright 2018 gRPC authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
namespace Grpc;

use Google\Protobuf\Internal\Message;
use Grpc\Internal\InterceptorChannel;

/**
 * CallInvoker is used to pass the self defined channel into the stub,
 * while intercept each RPC with the channel accessible.
 *
 * @template T of Message
 */
interface CallInvoker
{
    /**
     * @param string $hostname
     * @param array<string, mixed> $opts
     *
     * @return Channel|InterceptorChannel
     */
    public function createChannelFactory($hostname, $opts);

    /**
     * @param Channel|InterceptorChannel $channel
     * @param string $method
     * @param array{0: class-string<T>, 1: string} $deserialize
     * @param array<string, mixed> $options
     *
     * @return UnaryCall
     */
    public function UnaryCall($channel, $method, $deserialize, $options);

    /**
     * @param Channel|InterceptorChannel $channel
     * @param string $method
     * @param array{0: class-string<T>, 1: string} $deserialize
     * @param array<string, mixed> $options
     *
     * @return ClientStreamingCall
     */
    public function ClientStreamingCall($channel, $method, $deserialize, $options);

    /**
     * @param Channel|InterceptorChannel $channel
     * @param string $method
     * @param array{0: class-string<T>, 1: string} $deserialize
     * @param array<string, mixed> $options
     *
     * @return ServerStreamingCall
     */
    public function ServerStreamingCall($channel, $method, $deserialize, $options);

    /**
     * @param Channel|InterceptorChannel $channel
     * @param string $method
     * @param array{0: class-string<T>, 1: string} $deserialize
     * @param array<string, mixed> $options
     *
     * @return BidiStreamingCall
     */
    public function BidiStreamingCall($channel, $method, $deserialize, $options);
}

File: src/csharp/Grpc.Core.Api/CallInvoker.cs
#region Copyright notice and license

// Copyright 2015-2016 gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#endregion

using System.Threading.Tasks;

namespace Grpc.Core
{
    /// <summary>
    /// Abstraction of client-side RPC invocation.
    /// </summary>
    public abstract class CallInvoker
    {
        /// <summary>
        /// Invokes a simple remote call in a blocking fashion.
        /// </summary>
        public abstract TResponse BlockingUnaryCall<TRequest, TResponse>(Method<TRequest, TResponse> method, string host, CallOptions options, TRequest request)
            where TRequest : class
            where TResponse : class;

        /// <summary>
        /// Invokes a simple remote call asynchronously.
        /// </summary>
        public abstract AsyncUnaryCall<TResponse> AsyncUnaryCall<TRequest, TResponse>(Method<TRequest, TResponse> method, string host, CallOptions options, TRequest request)
            where TRequest : class
            where TResponse : class;

        /// <summary>
        /// Invokes a server streaming call asynchronously.
        /// In server streaming scenario, client sends on request and server responds with a stream of responses.
        /// </summary>
        public abstract AsyncServerStreamingCall<TResponse> AsyncServerStreamingCall<TRequest, TResponse>(Method<TRequest, TResponse> method, string host, CallOptions options, TRequest request)
            where TRequest : class
            where TResponse : class;

        /// <summary>
        /// Invokes a client streaming call asynchronously.
        /// In client streaming scenario, client sends a stream of requests and server responds with a single response.
        /// </summary>
        public abstract AsyncClientStreamingCall<TRequest, TResponse> AsyncClientStreamingCall<TRequest, TResponse>(Method<TRequest, TResponse> method, string host, CallOptions options)
            where TRequest : class
            where TResponse : class;

        /// <summary>
        /// Invokes a duplex streaming call asynchronously.
        /// In duplex streaming scenario, client sends a stream of requests and server responds with a stream of responses.
        /// The response stream is completely independent and both side can be sending messages at the same time.
        /// </summary>
        public abstract AsyncDuplexStreamingCall<TRequest, TResponse> AsyncDuplexStreamingCall<TRequest, TResponse>(Method<TRequest, TResponse> method, string host, CallOptions options)
            where TRequest : class
            where TResponse : class;
    }
}
